server/src/main/java/org/apache/iotdb/db/engine/compaction/writer/CrossSpaceCompactionWriter.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.db.engine.compaction.writer;
+
+import org.apache.iotdb.db.engine.storagegroup.TsFileResource;
+import org.apache.iotdb.tsfile.file.metadata.TimeseriesMetadata;
+import org.apache.iotdb.tsfile.read.TsFileSequenceReader;
+import org.apache.iotdb.tsfile.write.writer.RestorableTsFileIOWriter;
+import org.apache.iotdb.tsfile.write.writer.TsFileIOWriter;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class CrossSpaceCompactionWriter extends AbstractCompactionWriter {
+  // target fileIOWriters
+  private List<TsFileIOWriter> fileWriterList = new ArrayList<>();
+
+  // source tsfiles
+  private List<TsFileResource> seqTsFileResources;
+
+  private int seqFileIndex;
+
+  private final boolean[] isCurrentDeviceExist;
+
+  private final long[] currentDeviceEndTime;
+
+  private final boolean[] isEmptyFile;
+
+  public CrossSpaceCompactionWriter(
+      List<TsFileResource> targetResources, List<TsFileResource> seqFileResources)
+      throws IOException {
+    isCurrentDeviceExist = new boolean[seqFileResources.size()];
+    currentDeviceEndTime = new long[seqFileResources.size()];
+    isEmptyFile = new boolean[seqFileResources.size()];
+    for (int i = 0; i < targetResources.size(); i++) {
+      this.fileWriterList.add(new RestorableTsFileIOWriter(targetResources.get(i).getTsFile()));
+      isEmptyFile[i] = true;
+    }
+    this.seqTsFileResources = seqFileResources;
+    seqFileIndex = 0;
+  }
+
+  @Override
+  public void startChunkGroup(String deviceId, boolean isAlign) throws IOException {
+    this.deviceId = deviceId;
+    this.isAlign = isAlign;
+    checkIsDeviceExistAndGetDeviceEndTime();
+    for (; seqFileIndex < seqTsFileResources.size(); seqFileIndex++) {
+      if (isCurrentDeviceExist[seqFileIndex]) {
+        fileWriterList.get(seqFileIndex).startChunkGroup(deviceId);
+      }
+    }
```

Comment:
Or put the reset logic to end function, both ok, I prefer the second one, because you also reset it in `endMeasurement` function, but not reset it in `startMeasurement`.

Reply:
resolved